Source/WindowsShellExtension/dllmain.cpp

```diff
@@ -185,6 +185,63 @@ namespace {
         return hres;
     }
 
+    // Adapted from code generated with Google Gemini 2.0 Flash
+    _Check_return_ _Success_return_
+    bool ExtractUrlFromShortcut(_In_ const std::filesystem::path& filePath, _Out_ std::string& url) {
+        // Check if file exists and is readable
+        if (!std::filesystem::exists(filePath) || !std::filesystem::is_regular_file(filePath))
+            return false;
+
+        // Open the file in binary mode
+        std::ifstream file(filePath, std::ios::binary);
+
+        // Check if file opened successfully
+        if (!file.is_open())
+            return false;
+
+        // Read the entire file content
+        std::string content;
+        file.seekg(0, std::ios::end);
+        content.resize(file.tellg());
```

Comment:
There should be a security check that the file is not too big, e.g. leaving if file size is > 4096 bytes (a modern storage block size).